src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicConnection.cs

```diff
@@ -554,13 +557,34 @@ internal override ValueTask ConnectAsync(CancellationToken cancellationToken = d
             Debug.Assert(_state.StateGCHandle.IsAllocated);
 
             _state.Connection = this;
+            uint status;
+            if (_remoteEndPoint is IPEndPoint)
+            {
+                SOCKADDR_INET address = MsQuicAddressHelpers.IPEndPointToINet((IPEndPoint)_remoteEndPoint);
+                unsafe
+                {
+                    status = MsQuicApi.Api.SetParamDelegate(_state.Handle, QUIC_PARAM_LEVEL.CONNECTION, (uint)QUIC_PARAM_CONN.REMOTE_ADDRESS, (uint)sizeof(SOCKADDR_INET), (byte*)&address);
+                    QuicExceptionHelpers.ThrowIfFailed(status, "Failed to connect to peer.");
+                }
+
+                if (targetHost == null)
+                {
+                    targetHost = ((IPEndPoint)_remoteEndPoint).Address.ToString();
+                }
+            }
+            else if (targetHost == null)
```

Comment:
Not sure I understand this correctly... but it seems like this will result in the TargetHost from the clientAuthenticationOptions overriding the hostname from the DnsEndPoint. Is that really what we want?

Reply:
Similarly above for IPEndPoint...

Reply:
yes, that is what we do in SslStream. The SNI is used to identify particular virtual server on the same IP not the other way around. In HttpClient the TargetHost would be set either from `Host:` header or to Host from URI. What actual IP you connect to really irrelevant for TLS handshake. Since the clientAuthenticationOptions seems optional, I left the existing logic as fall-back.

Reply:
What I mean is this:

If I specify the RemoteEndPoint as a DnsEndPoint to "foo.com:1234", but I then set the TargetHost on ClientAuthenticationOptions to "bar.com", it seems like we will actually try to **connect** to "bar.com" rather than connect to "foo.com", which seems wrong. Similarly for an IPEndPoint.

With SslStream we specify the TargetHost separately from the address that we actually connect to. It seems like perhaps that is not possible using msquic? If so, I think we are best off just ignoring TargetHost here.

Reply:
yes, you are right and I think we should ignore TargetHost here as you suggested. I'll work with @nibanks to see if we can close the gap somehow.

Reply:
updated to always use the Host in case of DnsEndPoint.